### ISSUE TITLE
Wrong Marshalling when Entity is `Optional`

### ISSUE THREAD

**Issue by jlprat** [akka/akka-http#559]
When completing a route with a Java `Optional` the output is not the one I would expect:
## Expected Behavior
* If `Optional[T]` is empty: completed with an empty `Entity`
* If `Optional[T]` is not empty: completed with proper marshalling of `T`

## Actual Behavior
* If `Optional[T]` is empty: completed with `{"present": false}`
* If `Optional[T]` is not empty: completed with `{"present": true}`

I have a gist where one can reproduce this: https://gist.github.com/jlprat/5e1c27ce4b912e0a3eb0ad56237246d7

**Comment by rbudzko** [comment]
In theory (verified) it is enough to use brand new `2.8.5` and a primary problem disappears, so:

```
val jackson = "com.fasterxml.jackson.core"  % "jackson-databind" % "2.8.5"
val jacksonModules = "com.fasterxml.jackson.datatype" % "jackson-datatype-jdk8" % "2.8.5"
```

Thanks, to upgrading `ObjectMapper` can be configured to handle `Optional` in nicer way:

```
new ObjectMapper()
    .enable(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY)
    .registerModule(new Jdk8Module())
    .setSerializationInclusion(JsonInclude.Include.NON_ABSENT);
```

Unfortunatelly due to `OptionalSerializer` implementation we will end up with `"null"` instead of `""` for `Optional.empty()`. We would probably have to code sth around delegation so values are adapter before passing to scala directives - anyway it seems hackish...

**Comment by jlprat** [comment]
Would marshalling to null be acceptable? Or do would it be preferred to have the same behavior like in Scala (`None` -> "")?

**Comment by rbudzko** [comment]
I believe it is not. First of all the definition of emptiness would have to be different for java vs scala. I don't think it is OK, because confusing. Otherwise `rejectEmptyResponse` will not work because `""` is not a `"null"`. I think that adapter around marshaller could be a fixup, but it is still a non obvious solution. 

I know that you are still celebrating @ktoso 🎉 , but what is your view on this? It is kind of common problem and people like to transform `"null"` to `null` or `""` depending on their applications in java, but here we are talking about an unknown usage.

**Comment by jlprat** [comment]
I also would prefer to have the same exact behavior (returning ""). Let's see what's their opinion on this.

**Comment by ktoso** [comment]
Not still celebrating but working hard outlining our roadmap etc ;-) Will get back here later, thanks for your patience.

**Comment by ktoso** [comment]
It should be consistent with how Scala/Option is handled.
We can upgrade the dependency if it's compatible - is it? // Note to self, enable MiMa validation
`""` is a bit weird since it could be seen as `Some("")` - generally though we should make this consistent with what the scaladsl does now.

**Comment by rbudzko** [comment]
Yeah `""` is tricky, but it looks like jackson has same approach and produces `"null"` if serializing `null` or `Optional.empty()` as a **root** objects. Shall we check for `"null"` in `Jackson.toJson` and replace with `""`? Passes tests but sounds fishy. Is it acceptable in your opinion @ktoso ?

**Comment by jrudolph** [comment]
IMO having a generic `Option[T]` marshaller will always be problematic as it gives the illusion that we would handle `None` automagically for any content-type. Instead, there could be arbitrary representations of how `None` should be marshalled depending on the actual type `T` and the target content-type.

Wouldn't it make sense to remove the generic Option marshaller (compatibly somehow...) and fall back to more concrete marshallers for Option handling?

E.g. if you already have a json format defined for a `T`, you will also have a json format defined for `Option[T]`. So, for `application/json`, it will still work, producing a valid response.

WDYT?

**Comment by ktoso** [comment]
Yeah that sounds better.
The Option marshaller has been mostly confusing people so far...

I wonder how we can remove it. We'd need to deprecate etc now

**Comment by rbudzko** [comment]
Also sleazy hacks like in my PR would not be required. So, are we pivoting into removal of generic `Option` support from scala dsl rather then adding this possibility to java dsl?

**Comment by ktoso** [comment]
Closed the docs ticket, instead we should try fixing the directive.

@jrudolph's comment makes a lot of sense IMO, https://github.com/akka/akka-http/issues/559#issuecomment-265459534
